### ISSUE TITLE
Monorepo with multiple crates that can not share a workspace is not possible to build?

### ISSUE THREAD

**Issue by molysgaard** [nix-community/naersk#297]
Hi, I have this, arguably edge-case, monorepo with multiple rust crates.

Most of the crates are in the same workspace, meaning they build as one single whole and share dependencies.
But, some of the applications in the monorepo can not be in the workspace, this happens when several crates have different and incompatible native dependencies. This is a case cargo can not handle for a single workspace, and one needs to use the exclude option in the workspace Cargo.toml and build the conflicting apps as separate crates.
This is a bit tedious, but the end solution works well and integrates well with cargo, rust, rust-analyzer etc.

Now, I would like to use naersk to build one of these conflicting packages.
I have a minimal example that imitates the issue in this repo.
https://github.com/molysgaard/naersk-monorepo-multiple-crates
It should be only a `git clone && nix build` to test and see the issue.
What I am struggling with is telling naersk that path dependencies work on a superfolder of the `Cargo.toml` folder.

I would be very happy for any pointers to make this usecase work!

**Comment by Patryk27** [comment]
Ouch, yeah, something's off with the logic there; for the time being you can do:

```patch
diff --git a/flake.nix b/flake.nix
index 7856242..df67ae4 100644
--- a/flake.nix
+++ b/flake.nix
@@ -46,6 +46,7 @@
       root = ./app/myappb;
       gitSubmodules = true;
       buildInputs = deps;
+      singleStep = true;
     };
 
     devShells.x86_64-linux.default = pkgs.mkShell {
```

tl;dr by default, Naersk splits given project into two parts - the "vendored" code (i.e. external libraries) and "your" code; this allows Naersk to compile the vendored code just once and then if only your code changes, instead of re-building all of the crates, Naersk rebuilds only your application (so it's kinda like Cargo's incremental builds); and it seems there's something off with the logic in that part of Naersk - but fortunately, for the time being, you can simply turn that feature off 🙂 

I'll try to a look over the weekend and figure out what's wrong.

**Comment by molysgaard** [comment]
Thank you so much for a workaround! I would be glad to contribute if there is anything I can do to help :slightly_smiling_face:

**Comment by molysgaard** [comment]
@Patryk27 Hi. I did some tests, and it seems the workaround is not a workaround after all.

If add the `singleStep = true` to the build. The build succeeds, but it does not build the correct crate. If you run `nix build` and then `ls result` it only contains `myappa`, while it should contain `myappb`.

```
git clone https://github.com/molysgaard/naersk-monorepo-multiple-crates.git
cd molysgaard/naersk-monorepo-multiple-crates
# add the `singeStep = true;` line to `flake.nix`
nix build
ls result
```

Do you have any idea why this happens?

**Comment by Patryk27** [comment]
Oh, that's interesting - on the top of my head I'm not sure what might be wrong, but I'll take a look :smile:

**Comment by Patryk27** [comment]
Alright, this should do it 🙂 

```
packages.x86_64-linux.default = naersk'.buildPackage {
  src = ./.;
  root = ./.;
  gitSubmodules = true;
  buildInputs = deps;

  preBuild = ''
    cd app/myappb
  '';
};
```

**Comment by Patryk27** [comment]
Hi, to keep the issue tracker tidy, I'm going to close this issue - please feel free to open it again if you find something not working! 🙂

**Comment by matdibu** [comment]
I got it to work with something like this:
```nix
naersk'.buildPackage {
  src = <path to monorepo root>;
  root = <path to crate in in monorepo>;
  overrideMain = {
    preBuild = ''
      cd <monorepo-relative path to crate>
    '';
  };
}
```

**Comment by matdibu** [comment]
even simpler, this works

```nix
naersk'.buildPackage {
  src = ./.;
  override = {
    preBuild = ''
      cd app/myappb
    '';
  };
};
```

I've made the changes to your example repo

https://github.com/matdibu/naersk-monorepo-multiple-crates

```
$ nix build
$ tree ./result/
./result/
└── bin
    └── myappb
$ ./result/bin/myappb 
hello from myliba
hello from mylibb
```

**Comment by matdibu** [comment]
weirdly enough, only the first solution (using `overrideMain`) works on my monorepo
```
error: opening file '/nix/store/cpw49gq7zy2m68w4ib3zpq5d4dnsl5mi-source/Cargo.toml': No such file or directory
```

ah, it's because mine isn't using cargo workspaces
